cmds/core/wget/wget_test.go

```diff
@@ -40,131 +43,147 @@ func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer closer()
+
+	var tests = []struct {
+		name        string
+		url         string // in
+		wantContent string // out
+		outputPath  string
+		err         error
+	}{
+		{
+			name:        "ipv4",
+			url:         fmt.Sprintf("%s/200", srv.URL),
+			wantContent: content,
+			outputPath:  filepath.Join(tmpDir, "basic"),
+			err:         nil,
+		},
+		{
+			name:        "localhost",
+			url:         strings.Replace(srv.URL, "127.0.0.1", "localhost", 1) + "/200",
+			wantContent: content,
+			outputPath:  filepath.Join(tmpDir, "ipv4"),
+			err:         nil,
+		},
+		// TODO: CircleCI does not support ipv6
+		// {
+		// 	name:    "ipv6",
+		// 	flags:   []string{},
+		// 	url:     "http://[::1]:%[1]d/200",
+		// 	content: content,
+		// 	retCode: 0,
+		// },
+		{
+			name:       "redirect",
+			url:        fmt.Sprintf("%s/302", srv.URL),
+			outputPath: filepath.Join(tmpDir, "redirect"),
+			err:        nil,
+		},
+		{
+			name: "4xx error",
+			url:  fmt.Sprintf("%s/404", srv.URL),
+			err:  curl.ErrStatusNotOk,
+		},
+		{
+			name: "5xx error",
+			url:  fmt.Sprintf("%s/500", srv.URL),
+			err:  curl.ErrStatusNotOk,
+		},
+		{
+			name:       "connection closed",
+			url:        fmt.Sprintf("http://localhost:%d/200", freePort),
+			outputPath: filepath.Join(tmpDir, "connection-closed"),
+			err:        io.EOF,
+		},
+		{
+			name: "empty url",
+			url:  "",
+			err:  errEmptyURL,
+		},
+	}
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			// Change the working directory to a temporary directory, so we can
-			// delete the temporary files after the test runs.
-			tmpDir := t.TempDir()
-
-			fileName := filepath.Base(tt.url)
-
-			args := append(tt.flags,
-				"-O", filepath.Join(tmpDir, fileName),
-				fmt.Sprintf(tt.url, port, unusedPort))
-			cmd := testutil.Command(t, args...)
-			output, err := cmd.CombinedOutput()
+			err := newCommand(tt.outputPath, tt.url).run()
 
-			// Check return code.
-			if err := testutil.IsExitCode(err, tt.retCode); err != nil {
-				t.Errorf("exit code: %v, output: %s", err, string(output))
+			if tt.err == nil && err != nil {
+				t.Errorf("expected nil, got: %v", err)
+			}
+			if tt.err != nil && !errors.Is(err, tt.err) {
+				t.Errorf("expected: %v, got: %v", tt.err, err)
 			}
 
-			if tt.content != "" {
-				content, err := os.ReadFile(filepath.Join(tmpDir, fileName))
+			if tt.wantContent != "" {
+				content, err := os.ReadFile(tt.outputPath)
 				if err != nil {
-					t.Errorf("File %s was not created: %v", fileName, err)
+					t.Errorf("file %s was not created: %v", tt.outputPath, err)
 				}
 
 				// Check content.
-				if string(content) != tt.content {
-					t.Errorf("Want:\n%#v\nGot:\n%#v", tt.content, string(content))
+				if string(content) != tt.wantContent {
+					t.Errorf("wanted:\n%#v\ngot:\n%#v", tt.wantContent, string(content))
 				}
 			}
 		})
 	}
 }
 
-func TestMain(m *testing.M) {
-	testutil.Run(m, main)
+func TestDefaultOutputPath(t *testing.T) {
```

Comment:
Can this be integrated into TestWget?  Each test case already has `outputPath` to set the flag,  maybe you can add `wantOutputPath` and update the test case to read that file?       That would test the end-to-end functionality of wget.

TestDefaultOutputPath is testing an implementation detail - if one day in the future we wanted to rename the function or use some other way to determine the output path then we'd have to update this test code, even though the user's experience wouldn't change.   If you test this functionality as part of TestWget then we can change the implementation in the future and not have to update the tests.

Thanks!

Reply:
> index.html will depend from current directory and it don't work well with tempDir, so I don't think it worth to do.

I think this can be solved if you simply change into a tempdir: `os.Chdir(t.TempDir())`.   Here's a working example  that I came up with: https://gist.github.com/abrender/635144e451259825eb9e00527108205d

Reply:
Yes it can, but for one line of code that extracted and tested separately is't worth it? I mean I never made a chdir in test and it definitely will work here. But I'm afraid how it will affect other tests? Should I defer it back to original and what about defer and if we have Fatal? IMHO it's just became to complex.

Reply:
Go itself uses chdir while testing file / directory stuff (example: https://github.com/golang/go/blob/460fd63cccd2f1d16fc4b1b761545b1649e14e28/src/io/fs/walk_test.go#L102-L109).  Yes, good point about calling defer back to the original directory :)

I appreciate you questioning the trade off and complexity here. However I do believe that it's worth it here, ensuring that the correct output file gets written is a critical part of the program.  While testing `defaultOutputPath` is good, there's no guarantee that the function itself will be called or used correction - which is what this test covers.